core/input/input.cpp

```diff
@@ -870,6 +870,25 @@ void Input::parse_input_event(const Ref<InputEvent> &p_event) {
 
 	ERR_FAIL_COND(p_event.is_null());
 
+#ifdef DEBUG_ENABLED
+	uint64_t curr_frame = Engine::get_singleton()->get_process_frames();
+	if (p_event->_get_parsed_frame() == curr_frame) {
+		// It would be technically safe to send the same event in cases such as:
+		// - After an explicit flush.
+		// - In platforms using buffering when agile flushing is enabled, after one of the mid-frame flushes.
+		// - If platform doesn't use buffering and event accumulation is disabled.
+		// - If platform doesn't use buffering and the event type is not accumulable.
+		// However, it wouldn't be reasonable to ask users to remember the full ruleset and be aware at all times of the possibilites of the
+		// target platform, project settings and engine internals, which may change without notice.
+		// Therefore, the guideline is, "don't send the same event object more than once per frame".
+		WARN_PRINT(
+				"An input event of type " + p_event->get_class_name() + " is being buffered more than once in the same frame, which is unsafe.\n" +
+				"If you are generating events with scripting, you have to instantiate a new event instead of sending the same one more than once, unless the original one was sent at an earlier frame.");
+	}
+
+	p_event->_set_parsed_frame(curr_frame);
```

Comment:
Maybe this could use Object metadata instead of adding a debug-only public method in the API.

Reply:
Not sure about the tradeoffs. I guess @reduz can also give an opinion on this.

Reply:
I've done the change to use meta. The only caveat is that the public `parse_input_event()` function now has to take a non-const argument. I think that overall this is a better approach, though.

Reply:
Changing my mind again.

In order to allow casting `Ref<T> &` to `Ref<Base_of_T> &`, I had to add the following operator to `ref_counted.h`:
```C++
#pragma warning(push)
#pragma warning(disable : 4717) // MSVC thinks this causes infinite recursion.
	template <class T_Base,
			class = std::enable_if<std::is_base_of<T_Base, T>::value>::type>
	operator Ref<T_Base> &() & {
		return *this;
	}
#pragma warning(pop)
```

That works well, but causes compile errors in the intricate `PtrToArg` thing, which are not easy at all to solve.

So, I'm going back to the first version of this PR.

Reply:
By the way, the fact that this refernce conversion operator I tentatively added interacts with the `PtrToArg` makes me think there's a way to make all that much simpler by allowing more powerful type conversions. But that's another story.